File: showcase/integrations/ms-agent-dotnet/agent/InterruptAgent.cs
// @region[backend-interrupt-tool]
using System.Text.Json;
using Microsoft.Agents.AI;
using Microsoft.Extensions.AI;
using OpenAI;
using System.ClientModel;

// =================
// Interrupt Agent Factory
// =================
//
// Adaptation note: the Microsoft Agent Framework (.NET) does not have a
// LangGraph-equivalent `interrupt()` primitive that can pause execution
// mid-tool and resume with a caller-supplied value. The scheduling demos
// use a frontend-provided `schedule_meeting` tool; AG-UI forwards that tool
// definition to the model, then the client renders the picker and resolves
// the tool call with the user's selected slot.
//
// This factory reuses the existing SharedStateAgent pattern for
// consistency with the rest of the showcase, even though state-sync isn't
// the primary concern for interrupt demos. The agent's system prompt
// instructs it to always call `schedule_meeting` whenever the user asks
// to book a call or schedule a meeting.
public sealed class InterruptAgentFactory
{
    private const string DefaultOpenAiEndpoint = "https://models.inference.ai.azure.com";

    private readonly IConfiguration _configuration;
    private readonly OpenAIClient _openAiClient;
    private readonly ILogger _logger;
    private readonly ILoggerFactory _loggerFactory;
    private readonly JsonSerializerOptions _jsonSerializerOptions;

    public InterruptAgentFactory(IConfiguration configuration, ILoggerFactory loggerFactory, JsonSerializerOptions jsonSerializerOptions)
    {
        _configuration = configuration;
        _loggerFactory = loggerFactory;
        _logger = loggerFactory.CreateLogger<InterruptAgentFactory>();
        _jsonSerializerOptions = jsonSerializerOptions;

        var githubToken = _configuration["GitHubToken"]
            ?? throw new InvalidOperationException(
                "GitHubToken not found in configuration. " +
                "Please set it using: dotnet user-secrets set GitHubToken \"<your-token>\" " +
                "or get it using: gh auth token");

        var endpointEnv = Environment.GetEnvironmentVariable("OPENAI_BASE_URL");
        var endpoint = endpointEnv ?? DefaultOpenAiEndpoint;
        _logger.LogInformation(
            "InterruptAgentFactory using OpenAI endpoint: {Endpoint} (from OPENAI_BASE_URL: {HasEnv})",
            endpoint,
            !string.IsNullOrEmpty(endpointEnv));

        _openAiClient = new(
            new ApiKeyCredential(githubToken),
            AimockHeaderPolicy.CreateOpenAIClientOptions(endpoint));
    }

    // @region[backend-tool-call]
    public AIAgent CreateInterruptAgent()
    {
        var chatClient = _openAiClient.GetChatClient("gpt-4o-mini").AsIChatClient();

        // No backend fallback tool is registered. If the frontend tool is
        // missing, the demo should fail visibly instead of bypassing the
        // picker with a server-side response.
        var chatClientAgent = new ChatClientAgent(
            chatClient,
            name: "InterruptAgent",
            description: @"You are a scheduling assistant. Whenever the user asks you to book a call
or schedule a meeting, you MUST call the `schedule_meeting` tool. Pass a short `topic`
describing the purpose and `attendee` describing who the meeting is with. After the tool
returns, confirm briefly whether the meeting was scheduled and at what time, or that the
user cancelled.",
            tools: []);

        return new SharedStateAgent(chatClientAgent, _jsonSerializerOptions, _loggerFactory.CreateLogger<SharedStateAgent>());
    }
    // @endregion[backend-tool-call]

}
// @endregion[backend-interrupt-tool]

File: showcase/integrations/pydantic-ai/src/agents/interrupt_agent.py
"""PydanticAI scheduling agent -- interrupt-adapted.

This agent powers two demos (gen-ui-interrupt, interrupt-headless) that in the
LangGraph showcase rely on the native ``interrupt()`` primitive with
checkpoint/resume. PydanticAI does NOT have that primitive, so we adapt by
delegating the time-picker interaction to a **frontend tool** that the agent
calls by name (``schedule_meeting``). The frontend registers the tool via
``useFrontendTool`` with an async handler; that handler renders the interactive
picker, waits for the user to choose a slot (or cancel), and resolves the tool
call with the result. The backend only defines the system prompt and advertises
no local ``schedule_meeting`` implementation -- the agent's tool call is
satisfied entirely by the frontend.

See ``src/agents/hitl_in_chat_agent.py`` for the related ``book_call`` pattern
used by the HITL-in-chat demos in this package.
"""

from __future__ import annotations

from textwrap import dedent

from pydantic_ai import Agent
from pydantic_ai.models.openai import OpenAIResponsesModel


# @region[backend-tool-call]
SYSTEM_PROMPT = dedent(
    """
    You are a scheduling assistant. Whenever the user asks you to book a call
    or schedule a meeting, you MUST call the `schedule_meeting` tool. Pass a
    short `topic` describing the purpose of the meeting and, if known, an
    `attendee` describing who the meeting is with.

    The `schedule_meeting` tool is implemented on the client: it surfaces a
    time-picker UI to the user and returns the user's selection. After the
    tool returns, briefly confirm whether the meeting was scheduled and at
    what time, or note that the user cancelled. Do NOT ask for approval
    yourself -- always call the tool and let the picker handle the decision.

    Keep responses short and friendly. After you finish executing tools,
    always send a brief final assistant message summarizing what happened so
    the message persists.
    """.strip()
)


agent = Agent(
    model=OpenAIResponsesModel("gpt-4o-mini"),
    system_prompt=SYSTEM_PROMPT,
)
# @endregion[backend-tool-call]


__all__ = ["agent"]
